File: Scala/src/main/scala/parrot/Parrot.scala
package parrot

import parrot.ParrotType._

class Parrot(parrotType: ParrotType, numberOfCoconuts: Int, voltage: Double, isNailed: Boolean) {

  def getSpeed: Double = {
    parrotType match {
      case EUROPEAN =>
        return getBaseSpeed
      case AFRICAN =>
        return Math.max(0, getBaseSpeed - getLoadFactor * numberOfCoconuts)
      case NORWEGIAN_BLUE =>
        return if ((isNailed)) 0 else getBaseSpeed(voltage)
    }
    throw new RuntimeException("Should be unreachable")
  }

  private def getBaseSpeed(voltage: Double): Double = {
    return Math.min(24.0, voltage * getBaseSpeed)
  }

  private def getLoadFactor: Double = {
    return 9.0
  }

  private def getBaseSpeed: Double = {
    return 12.0
  }

}

File: CSharp/Parrot/Parrot.cs
using System;
using System.Collections.Generic;

namespace Parrot
{
    public class Parrot
    {
        private readonly bool _isNailed;
        private readonly int _numberOfCoconuts;
        private readonly ParrotTypeEnum _type;
        private readonly double _voltage;

        public Parrot(ParrotTypeEnum type, int numberOfCoconuts, double voltage, bool isNailed)
        {
            _type = type;
            _numberOfCoconuts = numberOfCoconuts;
            _voltage = voltage;
            _isNailed = isNailed;
        }

        public double GetSpeed()
        {
            switch (_type)
            {
                case ParrotTypeEnum.EUROPEAN:
                    return GetBaseSpeed();
                case ParrotTypeEnum.AFRICAN:
                    return Math.Max(0, GetBaseSpeed() - GetLoadFactor() * _numberOfCoconuts);
                case ParrotTypeEnum.NORWEGIAN_BLUE:
                    return _isNailed ? 0 : GetBaseSpeed(_voltage);
                default:
                    throw new ArgumentOutOfRangeException();
            }
        }

        private double GetBaseSpeed(double voltage)
        {
            return Math.Min(24.0, voltage * GetBaseSpeed());
        }

        private double GetLoadFactor()
        {
            return 9.0;
        }

        private double GetBaseSpeed()
        {
            return 12.0;
        }

        public string GetCry()
        {
            string value;
            switch (_type)
            {
                case ParrotTypeEnum.EUROPEAN:
                    value = "Sqoork!";
                    break;
                case ParrotTypeEnum.AFRICAN:
                    value = "Sqaark!";
                    break;
                case ParrotTypeEnum.NORWEGIAN_BLUE:
                    value = _voltage > 0 ? "Bzzzzzz" : "...";
                    break;
                default:
                    throw new ArgumentOutOfRangeException();
            }
            return value;
        }
    }
}